### ISSUE TITLE
Unable to run privileged containers within docker:dind

### ISSUE THREAD

**Issue by 7ail** [kata-containers/kata-containers#10298]
# Get your issue reviewed faster

To help us understand the problem more quickly, please do the following:

1. Run the `kata-collect-data.sh` script, which is installed as part of Kata Containers
   or `kata-containers.collect-data`, which is installed as part of the Kata Containers
   snapcraft package.
   ```sh
   $ sudo kata-collect-data.sh > /tmp/kata.log
   ```
   or
   ```sh
   $ sudo kata-containers.collect-data > /tmp/kata.log
   ```
1. Review the output file (`/tmp/kata.log`) to ensure it doesn't
   contain any private / sensitive information
1. Paste the *entire* contents of the file into this issue as a comment
   (the script generates markdown format output).

The information provided will help us to understand the problem more quickly
so saves time for both of us! :smile:

# Description of problem
We believe that Kata 3.3 introduced a regression when we try to run a privileged container within a docker:dind container

More specifically changes introduced in this PR: https://github.com/kata-containers/kata-containers/pull/7531

# Expected result

We should be able to spin up a healthy privileged container within docker:dind

```
docker run --rm --privileged ubuntu:20.04 echo hello world

hello world
```

# Actual result

container initialisation fails with the following error

```
docker run --rm --privileged ubuntu:20.04 echo hello world


docker: Error response from daemon: failed to create task for container: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: error setting cgroup config for procHooks process: failed to write "a *:* rwm": write /sys/fs/cgroup/devices/docker/db123927e188abe1bc1791a020f11474c8d634fd39dbf425fd8e93dac8a757b5/devices.allow: operation not permitted: unknown.
```

# Further information

# Kata Containers survey

Please consider taking the survey to help us help you: https://openinfrafoundation.formstack.com/forms/kata_containers_user_survey

**Comment by justxuewei** [comment]
Hi @7ail. Thanks for reporting this. We are going to have a national holiday in the next 7 days. Please ping me if I forget to investigate this issue.

**Comment by 7ail** [comment]
Hey @justxuewei, just pinging you as a reminder.

**Comment by justxuewei** [comment]
@7ail Sorry for the late reply. I saw you mentioned "dind". Would you mind providing some additional information on how you start a Kata container?

**Comment by 7ail** [comment]
Hi @justxuewei 

For our use case, we are using the Kata runtime to spin up a docker:Dind container akin to

```
nerdctl run \
  --privileged \
  --security-opt=privileged-without-host-devices \
  --snapshotter devmapper \
  --runtime io.containerd.run.kata.v2 \
  -v /var/lib/docker:/var/lib/docker \
  -t docker:dind --debug --log-level=DEBUG
```

To reproduce the issue, we then ssh into the container
```
nerdctl exec -it service-docker sh
```

once inside we run 
```
docker run --rm --privileged ubuntu:20.04 echo hello world
```

which resulted in
```
docker: Error response from daemon: failed to create task for container: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: error setting cgroup config for procHooks process: failed to write "a *:* rwm": write /sys/fs/cgroup/devices/docker/db123927e188abe1bc1791a020f11474c8d634fd39dbf425fd8e93dac8a757b5/devices.allow: operation not permitted: unknown.
```

**Comment by justxuewei** [comment]
Hi @7ail, I have some updates for you.

TL;DR

You should enable both `privileged-without-host-devices` and `privileged_without_host_devices_all_devices_allowed`.

---

Start a container with `--privileged`, and then dump the OCI spec:

```
LinuxResources { devices: S
ome([LinuxDeviceCgroup { allow: false, typ: Some(A), major: None, minor: None, access: Some(\\\"rwm\\\") }, LinuxDeviceCgroup { allow: true, typ: Some
(C), major: Some(1), minor: Some(3), access: Some(\\\"rwm\\\") }, LinuxDeviceCgroup { allow: true, typ: Some(C), major: Some(1), minor: Some(8), acces
s: Some(\\\"rwm\\\") }, LinuxDeviceCgroup { allow: true, typ: Some(C), major: Some(1), minor: Some(7), access: Some(\\\"rwm\\\") }, LinuxDeviceCgroup
{ allow: true, typ: Some(C), major: Some(5), minor: None, access: Some(\\\"rwm\\\") }, LinuxDeviceCgroup { allow: true, typ: Some(C), major: Some(1),
minor: Some(5), access: Some(\\\"rwm\\\") }, LinuxDeviceCgroup { allow: true, typ: Some(C), major: Some(1), minor: Some(9), access: Some(\\\"rwm\\\")
}, LinuxDeviceCgroup { allow: true, typ: Some(C), major: Some(5), minor: Some(1), access: Some(\\\"rwm\\\") }, LinuxDeviceCgroup { allow: true, typ: S
ome(C), major: Some(136), minor: None, access: Some(\\\"rwm\\\") }, LinuxDeviceCgroup { allow: true, typ: Some(C), major: Some(5), minor: Some(2), acc
ess: Some(\\\"rwm\\\") }]), ... }
```

where `LinuxDeviceCgroup { allow: false, typ: Some(A), major: None, minor: None, access: Some(\\\"rwm\\\") }` doesn't enable all device allowlisting.

Containerd community has already discussed and fixed the same issue for Kata Containers, but I have no idea about if it be worked with nerdctl:

- https://github.com/containerd/containerd/pull/5686
- https://github.com/containerd/containerd/issues/5679
- https://github.com/containerd/containerd/issues/6643
